aioshelly/wsrpc.py

```diff
@@ -0,0 +1,194 @@
+"""WsRpc for Shelly."""
+from __future__ import annotations
+
+import asyncio
+import logging
+from dataclasses import dataclass
+from datetime import datetime
+from typing import Any, Dict
+
+from aiohttp import ClientWebSocketResponse, client_exceptions
+
+from .const import WS_HEARTBEAT_SEC
+from .exceptions import (
+    CannotConnect,
+    InvalidMessage,
+    JSONRPCError,
+    RPCError,
+    RPCTimeout,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+
+@dataclass
+class RPCCall:
+    """RPCCall class."""
+
+    call_id: int
+    method: str
+    params: Dict[str, Any] | None
+    src: str | None = None
+    dst: str | None = None
+    sent_at: datetime | None = None
+    resolve: asyncio.Future | None = None
+
+    @property
+    def request_frame(self):
+        """Request frame."""
+        msg = {
+            "id": self.call_id,
+            "method": self.method,
+            "src": self.src,
+        }
+        for obj in ("params", "dst"):
+            if getattr(self, obj) is not None:
+                msg[obj] = getattr(self, obj)
+        return msg
+
+
+class WsRPC:
+    """WsRPC class."""
+
+    def __init__(self, ip_address: str, on_notification):
+        """Initialize WsRPC class."""
+        self._ip_address = ip_address
+        self._on_notification = on_notification
+        self._rx_task = None
+        self._websocket: ClientWebSocketResponse | None = None
+        self._calls: Dict[str, str] = {}
+        self._call_id = 1
+        self._src = f"aios-{id(self)}"
+        self._dst = None
+
+    @property
+    def _next_id(self):
+        self._call_id += 1
+        return self._call_id
+
+    async def connect(self, aiohttp_session):
```

Comment:
If you want, you could simplify the logic of this class if you just make `connect` do all the work and not create a `self._rx_task`. That way the disconnect event is built-in -> the method finishes. It's what we do for ZJS https://github.com/home-assistant-libs/zwave-js-server-python/blob/master/zwave_js_server/client.py#L162  

You could even make it go one up and make it an async iterator that iterates over events and then in the end just finishes when connection done. We do that in aiohue: https://github.com/home-assistant-libs/aiohue/blob/master/aiohue/clip.py#L11 and consume it [like this](https://github.com/home-assistant-libs/aiohue/blob/master/aiohue/bridge.py#L162)

Reply:
I think this changes too much the current implementation and the benefit from this is small so let's leave that to a future change if we make it